### PR TITLE
adding dynamically_quantize_per_tensor

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #24
* #22
* #21
* #20
* __->__ #18

Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

### CODE REVIEW

**Comment by vkuzo** [comment]
> largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

can we spell these out?  This is really important to understand and document.

File: ao_experimental/quant_primitives.py [hunk 1]
```diff
@@ -2,6 +2,60 @@
...
+    # get min and max
+    # min_val, max_val = torch.aminmax(x) # compiled triton code is the same for min/max and aminmax
+    min_val = torch.min(x)
+    max_val = torch.max(x)
+
+    min_val_neg = torch.min(min_val, torch.zeros_like(min_val))
+    max_val_pos = torch.max(max_val, torch.zeros_like(max_val))
+
+    # calculate scale and zero point based on min and max
+    # reference: https://github.com/pytorch/pytorch/blob/e779a30d5097714acea011da6a554e43810b5d0e/aten/src/ATen/native/quantized/cpu/QuantUtils.h#L107
```
Comment (vkuzo) [inline ao_experimental/quant_primitives.py:41]: nit: would be good to add comments about specific precision choices being made here.

**Review by vkuzo — APPROVED** [review]

**Comment by HDCharles** [comment]
> 

done

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

File: ao_experimental/quant_primitives.py (hunk 1 above)
Comment (HDCharles) [reply ao_experimental/quant_primitives.py:41]: done

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

**Pull request closed without merging**